### DISCUSSION TITLE
g3proxy生成版本发布包时报错：failed to download `cmake v0.1.50`

### DISCUSSION THREAD

**Discussion by pyfx1010** [bytedance/g3#124] [General]
请问，我在centos7和9上执行命令：./scripts/release/build_tarball.sh g3proxy-v1.7.25
生成版本发布包时均卡在如下报错处无法进行:：

==> adding vendor code from cargo
error: failed to sync

Caused by:
  failed to download packages

Caused by:
  failed to download `cmake v0.1.50`

Caused by:
  attempting to make an HTTP request, but --frozen was specified

我在cmake官网上能找到的最低版本是v1.2，无法找到v0.1.50相关版本，网上也搜索不到相关资料，请指导一下如何解决，谢谢

**Comment by zh-jq-b** [comment]
拉一下最新代码执行下

**Comment by pyfx1010** [comment]
重新拉代码，当前这个报错解决了，在centos7下，再次执行：./scripts/release/build_tarball.sh g3proxy-v1.7.25
仍然有报错，并且未产生g3proxy-1.7.25.tar.xz包，错误信息如下：

The HTML pages are in g3proxy/doc/_build/html.
==> moving package files
==> building final tarball
tar: g3\:1000: Invalid owner
tar: Error is not recoverable: exiting now
clear temp build dir /tmp/build.g3proxy-1.7.25.LaZ

请帮忙解决一下，感谢

**Reply by zh-jq-b** [reply]
这个需要自行适配下老版本的tar，删除不支持的参数

**Reply by pyfx1010** [reply]
谢谢，删除build_tarball.sh中打包命令中的 ${REPRODUCIBLE_OPTS} 参数可以执行成功了，下一步执行：rpmbuild -ta g3proxy-1.7.25.tar.xz 时又有报错，内容哪下，请再次协助：
  running cd "/root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/build/src" && AR="ar" CC="cc" RANLIB="ranlib" "perl" "./Configure" "--prefix=/root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/install" "--openssldir=/usr/local/ssl" "no-dso" "no-shared" "no-ssl3" "no-tests" "no-comp" "no-zlib" "no-zlib-dynamic" "--libdir=lib" "no-md2" "no-rc5" "no-weak-ssl-ciphers" "no-camellia" "no-idea" "no-seed" "linux-x86_64" "-O2" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64"

  --- stderr
  Can't locate IPC/Cmd.pm in @INC (@INC contains: /root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/build/src/util/perl /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 . /root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/build/src/external/perl/Text-Template-1.56/lib) at /root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/build/src/util/perl/OpenSSL/config.pm line 19.
  BEGIN failed--compilation aborted at /root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/build/src/util/perl/OpenSSL/config.pm line 19.
  Compilation failed in require at ./Configure line 23.
  BEGIN failed--compilation aborted at ./Configure line 23.
  thread 'main' panicked at '

  Error configuring OpenSSL build:
      Command: cd "/root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/build/src" && AR="ar" CC="cc" RANLIB="ranlib" "perl" "./Configure" "--prefix=/root/rpmbuild/BUILD/g3proxy-1.7.25/target/release-lto/build/tongsuo-sys-d8b73009b39a50f8/out/openssl-build/install" "--openssldir=/usr/local/ssl" "no-dso" "no-shared" "no-ssl3" "no-tests" "no-comp" "no-zlib" "no-zlib-dynamic" "--libdir=lib" "no-md2" "no-rc5" "no-weak-ssl-ciphers" "no-camellia" "no-idea" "no-seed" "linux-x86_64" "-O2" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64"
      Exit status: exit status: 2

      ', /root/rpmbuild/BUILD/g3proxy-1.7.25/vendor/openssl-src/src/lib.rs:573:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: Bad exit status from /var/tmp/rpm-tmp.jbhzUc (%build)

RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.jbhzUc (%build)

**Reply by zh-jq-b** [reply]
yum install perl-IPC-Cmd

**Reply by pyfx1010** [reply]
你好，上面的问题解决了，但又有新的报错，前面Compiling处没有处理geoip，后面install处却需要用到geoip，提示找不到

具体报错信息如下：
   Compiling g3proxy-ctl v0.1.0 (/root/rpmbuild/BUILD/g3proxy-1.7.25/g3proxy/utils/ctl)
   Compiling g3proxy-ftp v0.1.0 (/root/rpmbuild/BUILD/g3proxy-1.7.25/g3proxy/utils/ftp)
    Finished release-lto [optimized] target(s) in 34m 22s
+ sh g3proxy/service/generate_systemd.sh
+ exit 0
Executing(%install): /bin/sh -e /var/tmp/rpm-tmp.5Crtyt
+ umask 022
+ cd /root/rpmbuild/BUILD
+ '[' /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64 '!=' / ']'
+ rm -rf /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64
++ dirname /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64
+ mkdir -p /root/rpmbuild/BUILDROOT
+ mkdir /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64
+ cd g3proxy-1.7.25
+ rm -rf /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64
+ install -m 755 -D target/release-lto/g3proxy /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64/usr/bin/g3proxy
+ install -m 755 -D target/release-lto/g3proxy-ctl /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64/usr/bin/g3proxy-ctl
+ install -m 755 -D target/release-lto/g3proxy-ftp /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64/usr/bin/g3proxy-ftp
+ install -m 755 -D target/release-lto/g3proxy-lua /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64/usr/bin/g3proxy-lua
+ install -m 755 -D target/release-lto/g3proxy-geoip /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64/usr/bin/g3proxy-geoip
install: cannot stat 'target/release-lto/g3proxy-geoip': No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.5Crtyt (%install)

**Reply by zh-jq-b** [reply]
拉一下最新代码的spec文件，忘了加feature了

**Reply by pyfx1010** [reply]
另外，每次rpmbuild都会在最后卡住很久，是编译的主机性能较低的原因吗？

   Compiling g3-json v0.3.0 (/root/rpmbuild/BUILD/g3proxy-1.7.25/lib/g3-json)
   Compiling g3proxy-ctl v0.1.0 (/root/rpmbuild/BUILD/g3proxy-1.7.25/g3proxy/utils/ctl)
   Compiling g3proxy-ftp v0.1.0 (/root/rpmbuild/BUILD/g3proxy-1.7.25/g3proxy/utils/ftp)
    Building [=======================> ] 310/311: g3proxy(bin)

**Reply by zh-jq-b** [reply]
lto链接优化比较慢

**Reply by pyfx1010** [reply]
我重新拉取了全部代码，rpmbuild时仍然报找不到geoip，请帮忙

+ install -m 755 -D target/release-lto/g3proxy-geoip /root/rpmbuild/BUILDROOT/g3proxy-1.7.25-1.el7.x86_64/usr/bin/g3proxy-geoip
install: cannot stat 'target/release-lto/g3proxy-geoip': No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.vHjtRa (%install)

**Reply by zh-jq-b** [reply]
用最新tag 1.7.26
